lib/srv/db/snowflake/models.go

```diff
@@ -0,0 +1,201 @@
+ limitations under the License.
+
+
+*/
+
+package snowflake
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+// loginRequest is the payload sent to /queries/v1/query-request endpoint.
+type loginRequest struct {
+	Data struct {
+		ClientAppID             string          `json:"CLIENT_APP_ID"`
+		ClientAppVersion        string          `json:"CLIENT_APP_VERSION"`
+		SvnRevision             string          `json:"SVN_REVISION"`
+		AccountName             string          `json:"ACCOUNT_NAME"`
+		LoginName               string          `json:"LOGIN_NAME,omitempty"`
+		Password                string          `json:"PASSWORD,omitempty"`
+		RawSAMLResponse         string          `json:"RAW_SAML_RESPONSE,omitempty"`
+		ExtAuthnDuoMethod       string          `json:"EXT_AUTHN_DUO_METHOD,omitempty"`
+		Passcode                string          `json:"PASSCODE,omitempty"`
+		Authenticator           string          `json:"AUTHENTICATOR,omitempty"`
+		SessionParameters       json.RawMessage `json:"SESSION_PARAMETERS,omitempty"`
+		ClientEnvironment       json.RawMessage `json:"CLIENT_ENVIRONMENT"`
+		BrowserModeRedirectPort string          `json:"BROWSER_MODE_REDIRECT_PORT,omitempty"`
+		ProofKey                string          `json:"PROOF_KEY,omitempty"`
+		Token                   string          `json:"TOKEN,omitempty"`
+	} `json:"data"`
+}
+
+// loginResponse is the payload returned by the /queries/v1/query-request endpoint.
+type loginResponse struct {
+	// use map here to not remove any fields when marshaling back.
+	Data map[string]interface{} `json:"data"`
+
+	Code    interface{} `json:"code"`
+	Message interface{} `json:"message"`
+	Success bool        `json:"success"`
+}
+
+// decodeLoginResponse decodes the bodyBytes as loginResponse struct. It uses json.Number to decode number.
+func decodeLoginResponse(bodyBytes []byte) (*loginResponse, error) {
+	loginResp := &loginResponse{}
+	decoder := json.NewDecoder(bytes.NewReader(bodyBytes))
+	// Use numbers to not modify the numbers values when processing payload.
+	decoder.UseNumber()
+	if err := decoder.Decode(loginResp); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return loginResp, nil
+}
+
+func (l *loginResponse) getTokens() (sessionTokens, error) {
+	getField := func(name string) (string, error) {
+		dataToken, found := l.Data[name]
+		if !found {
+			return "", trace.Errorf("Snowflake login response doesn't contain expected %s field", name)
+		}
+
+		token, ok := dataToken.(string)
+		if !ok {
+			return "", trace.Errorf("%s field returned by Snowflake API expected to be a string, got %T", name, dataToken)
+		}
+
+		return token, nil
+	}
+
+	getFieldInt := func(name string) (int64, error) {
+		dataToken, found := l.Data[name]
+		if !found {
+			return 0, trace.Errorf("Snowflake login response doesn't contain expected %s field", name)
+		}
+
+		validFor, ok := dataToken.(json.Number)
```

Comment:
nit: rename `validFor`

not suggesting changing the unmarshal logic here now, but i think some json marshaler/unmarshaler trick may also work https://go.dev/play/p/KzsG583PnV-

Reply:
Fixed, thanks for the suggestion: https://github.com/gravitational/teleport/pull/12564/commits/496d67e7e99b1f6bea5f4aedaa99d3bbc50d6ded